server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java

```diff
@@ -397,6 +397,32 @@ record StartRecoveryRequestToSend(StartRecoveryRequest startRecoveryRequest, Str
                     }
                     indexShard.recoverLocallyUpToGlobalCheckpoint(ActionListener.assertOnce(l));
                 })
+                // peer recovery can consume a lot of disk space, so it's worth cleaning up locally ahead of the attempt
+                // operation runs only if the previous operation succeeded, and returns the previous operation's result.
+                // Failures at this stage aren't fatal, we can attempt to recover and then clean up again at the end. #104473
+                .andThenApply(startingSeqNo -> {
+                    Store store = indexShard.store();
+                    store.incRef();
+                    try {
+                        store.cleanupAndVerify("cleanup before peer recovery", indexShard.snapshotStoreMetadata());
+                    } catch (IOException e) {
+                        // if we can't read a snapshot, then we don't have enough intact data locally to create one.
+                        // Make an attempt to clean up temporary files from the previous run. We could also just delete
+                        // everything in the directory but in that case we might want to be a little more thoughtful about
+                        // the exact exception triggered here. Corrupt indices might be recoverable without a full transfer, for
+                        // example.
+                        logger.info("cleanup before peer recovery failed on shard [{}]; exception: [{}]", indexShard.shardId(), e);
```

Comment:
logger.info(() -> format("cleanup before peer recovery failed on shard [%s]", indexShard.shardId()), e);

Reply:
We prefer to use the logger method which takes the exception. Also, `Supplier` is an option. But you can also pass `formatted string` directly. It's extremely common to use supplier for trace/debug logs which are normally disabled. Since warn, error, and info are normally enabled I don't think it makes a huge difference.

Reply:
I see. I was trying to be consistent with the [neighboring code](https://github.com/elastic/elasticsearch/blob/d6f8de22d697cd4db7a6f0eb179d6835867e9d35/server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java#L364) but I can see that it is [in](https://github.com/elastic/elasticsearch/blob/d6f8de22d697cd4db7a6f0eb179d6835867e9d35/server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java#L261) [flux](https://github.com/elastic/elasticsearch/blob/d6f8de22d697cd4db7a6f0eb179d6835867e9d35/server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java#L256) and will use the lambda form, and not eagerly stringify the exception.